xarray/core/dataset.py

```diff
@@ -5914,5 +5921,169 @@ def pad(
 
         return self._replace_vars_and_dims(variables)
 
+    def idxmin(
+        self,
+        dim: Hashable = None,
+        axis: int = None,
+        skipna: bool = None,
+        promote: bool = None,
```

Comment:
And generally I don't think we've given a `promote` option but, again, no harm in including

Reply:
This was the part I was most uncertain about, both the presence of the argument and its default.  I thought this was a bit of a special case since the data being tested is not the same as the data being returned, which can lead to non-intuitive results.

The options I thought of were:

   1. Always raisr an exception on operations that would insert an NaN.  This is the safest option, and the same as what `argmin` and `argmax` do, but would prevent an enormous number of valid operations and is overall not a good option.  It is the easiest to implement, though.
   2. Raise an exception if it would insert an NaN value in a data type that doesn't support it.  This is safe, but can have some unintuitive results since valid data for analysis can lead to invalid results.
   3. Only promote if NaN values are actually encountered.  This is safe, but makes the results hard to predict since they data type of the result depends on the specific values in the input.
   4. Always promote.  This works in the widest variety of situations, but can change data types where it isn't wanted.
  5. Ask

I settled on 5 with a default of 4 because I thought it worked most generally while still allowing people to restrict it only to save operations.

Reply:
Cool, I think that's very reasonable, and well laid-out options. 

Any other thoughts? @pydata/xarray on the general policy? I know we've had discussions on this before.

Reply:
I think (4) is a reasonable default choice. One other reason why it is nice is that Dask won't work great with the other options, because that would require doing a computational pass over the data to figure out the dtype.

Reply:
It occurs to me that there is at least one other possibly defensible behavior in the case of all NaN values: returning an arbitrarily selected coordinate value. This would still satisfy the invariant `array.sel({dim: ds.idxmin(dim)}) == array.min(dim)`.

Reply:
@shoyer I should have mentioned that approach.  I didn't include that approach since it would be very easy to do manually at the end.  But it wouldn't tell anyone that they are dealing with an unusual situation.

Reply:
How could you do "return an arbitrary value" without type promotion as a post-processing step?

Reply:
@shoyer It would still be promoted.  Trying to let people specify an arbitrary value would considerably increase the complexity of the code.  There would need to be an additional argument, and more complicated promotion handling.  As it stands now, the automatic promotion is decided right at the beginning, so we only need to handle two cases for the most part.  Adding a fourth option would make that more complicated, and would require additional checks for the argument, and possibly still require promotion if the dtype of the fill value is different than the coordinate dtype.  It just opens up an enormous can of worms and corner cases that need to be resolved and tested that I think aren't really worth it.

Reply:
Just to throw out another API option: what about having a `fill_value` argument instead of `promote`? The default (`fill_value=dtypes.NA`) would do type promotion for integer dtypes and always fill with NA. Other values (e.g., `fill_value=0`) could be used to avoid type promotion with an integer coordinate.

Advantages:

- No special cases to keep track of.
- Consistent with other xarray methods that take a `fill_value` argument.

Disadvantages:

- No built-in way to raise an error instead of promotion (but users could do this themselves pretty easily)
- No built-in way to "only promote if necessary" (but this is a weird non-type stable API that doesn't work great with Dask, anyways)

Reply:
That could work.

The corner case we would need to decide on is again promotion.

What happens if the fill value is a "higher" type in the numerical tower than
the original type?  What if it is lower?

   1. We could try to always convert to the fill dtype (or more often the
dtype equivalent to the Python native type), and raise and exception of it
doesn't work.
   2. We could promote the fill value or original data, whichever is
"lower".

What if someone tries to use a string type for numeric data or vice
versus?  If we do option 1 that is easy.  Otherwise we probably need to use
numpy casting rules?

What about an object dtype fill value?

What about a date/time related dtypes?

We would need to figure out how to handle these cases before I try to implement too much of this.

Reply:
I _think_ the right answer is to cast to the common dtype between the array and `fill_value`. This is what should automatically happen in xarray when you combine arrays of different types, e.g., using `duck_array_ops.where`:
https://github.com/pydata/xarray/blob/6378a711d50ba7f1ba9b2a451d4d1f5e1fb37353/xarray/core/duck_array_ops.py#L246

Reply:
Okay, I have done that.

Reply:
This is done, please check the lastest  commit.